contract_fixed_discount/models/abstract_contract_line.py

```diff
@@ -0,0 +1,47 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ContractAbstractContractLine(models.AbstractModel):
+    _inherit = "contract.abstract.contract.line"
+
+    discount_fixed = fields.Float(
+        string="Discount (Fixed)",
+        digits="Product Price",
+        default=0.00,
+        help="Fixed amount discount.",
+    )
+
+    @api.onchange("discount")
+    def _onchange_discount(self):
+        if self.discount:
+            self.discount_fixed = 0.0
+
+    @api.onchange("discount_fixed")
+    def _onchange_discount_fixed(self):
+        if self.discount_fixed:
+            self.discount = 0.0
+
+    @api.constrains("discount", "discount_fixed")
+    def _check_only_one_discount(self):
+        for rec in self:
+            for line in rec:
+                if line.discount and line.discount_fixed:
+                    raise ValidationError(
+                        _("You can only set one type of discount per line.")
+                    )
+
+    def _get_price_subtotal(self):
+        self.ensure_one()
+        subtotal = self.quantity * self.price_unit
+        if self.discount:
+            subtotal = super()._get_price_subtotal()
+        elif self.discount_fixed:
+            subtotal -= self.discount_fixed
+        return subtotal
```

Comment:
A quick redesign suggestion to allow a little bit more of flexibility
```suggestion
    def _get_discounted_price_subtotal(self):
        self.ensure_one()
        if self.discount_fixed:
            return self._get_price_subtotal() - self.discount_fixed
        return super()._get_discounted_price_subtotal()
```